src/work/Work.cpp

```diff
@@ -4,434 +4,229 @@
+                       });
 }
 
-void
-Work::onStart()
+bool
+Work::allChildrenDone() const
 {
+    return std::all_of(
+        mChildren.begin(), mChildren.end(),
+        [](std::shared_ptr<BasicWork> const& w) { return w->isDone(); });
 }
 
-void
-Work::onRun()
+bool
+Work::anyChildRunning() const
 {
-    scheduleSuccess();
+    return std::any_of(mChildren.begin(), mChildren.end(),
+                       [](std::shared_ptr<BasicWork> const& w) {
+                           return w->getState() ==
+                                  BasicWork::State::WORK_RUNNING;
+                       });
 }
 
-Work::State
-Work::onSuccess()
+bool
+Work::hasChildren() const
 {
-    return WORK_SUCCESS;
+    return !mChildren.empty();
 }
 
-void
-Work::onFailureRetry()
+bool
+Work::anyChildRaiseFailure() const
 {
+    return std::any_of(mChildren.begin(), mChildren.end(),
+                       [](std::shared_ptr<BasicWork> const& w) {
+                           return w->getState() ==
+                                  BasicWork::State::WORK_FAILURE;
+                       });
 }
 
-void
-Work::onFailureRaise()
+std::shared_ptr<BasicWork>
+Work::yieldNextRunningChild()
 {
-}
+    while (mNextChild != mChildren.end())
+    {
+        auto next = mNextChild;
+        mNextChild++;
+        assert(*next);
+        auto state = (*next)->getState();
+        if (state == State::WORK_RUNNING)
+        {
+            return *next;
+        }
+        else if (state == State::WORK_SUCCESS)
+        {
+            mNextChild = mChildren.erase(next);
+        }
+    }
 
-Work::State
-Work::getState() const
-{
-    return mState;
+    mNextChild = mChildren.begin();
+
+    return nullptr;
 }
 
-bool
-Work::isDone() const
+WorkSequence::WorkSequence(Application& app, std::string name,
+                           std::vector<std::shared_ptr<BasicWork>> sequence,
+                           size_t maxRetries)
+    : Work(app, std::move(name), maxRetries)
+    , mSequenceOfWork(sequence)
+    , mNextInSequence(mSequenceOfWork.begin())
 {
-    return mState == WORK_SUCCESS || mState == WORK_FAILURE_RAISE ||
-           mState == WORK_FAILURE_FATAL;
 }
 
-void
-Work::setState(Work::State st)
+BasicWork::State
+WorkSequence::doWork()
 {
-    auto maxR = getMaxRetries();
-    if (st == WORK_FAILURE_RETRY && (mRetries >= maxR))
+    if (mNextInSequence == mSequenceOfWork.end())
     {
-        CLOG(WARNING, "Work")
-            << "Reached retry limit " << maxR << " for " << getUniqueName();
-        st = WORK_FAILURE_RAISE;
+        // Completed all the work
+        assert(!hasChildren());
+        return State::WORK_SUCCESS;
     }
 
-    if (st != mState)
+    auto w = *mNextInSequence;
+    assert(w);
+    if (!hasChildren())
     {
-        CLOG(DEBUG, "Work") << "work " << getUniqueName() << " : "
-                            << stateName(mState) << " -> " << stateName(st);
-        mState = st;
+        w->startWork(wakeUpCallback());
+        addChild(w);
```

Comment:
This looks very strange to me. So WorkSequence adds current work item as a child, when it is done it clears all children and does the same with next item?

I wonder if derivering WorkSequence directly from BasicWork would not produce simpler code.

Reply:
That is a good point! `WorkSequence` relies on Work's functionality that ensures `crankWork` is called on children and on self in proper order; i.e. WorkSequence is a special case of Work, as it only has 1 child at a time. However, I do think you are right that it doesn't necessarily have to rely on Work. Instead, with the minimal code change, it could derive from BasicWork and dispatch crank calls to the child appropriately itself. The reason I think it's nice is because it will  cause `WorkSequence` to be scheduled less often - as `WorkSequence` doesn't do any actual work, but rather forwards the cranking call to an appropriate work.